src/bosh_azure_cpi/lib/cloud/azure/restapi/azure_client.rb

```diff
@@ -123,10 +123,44 @@ def rest_api_url(resource_provider, resource_type, resource_group_name: nil, nam
       url
     end
 
+    # get single resource
+    # example: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkInterfaces/{networkInterfaceName}
     def get_resource_by_id(url, params = {})
       result = nil
       begin
-        result = http_get(url, params)
+        uri = http_url(url, params)
+        response = http_get(uri)
+        result = JSON.parse(response.body) unless response.body.nil?
+      rescue AzureNotFoundError => e
+        result = nil
+      end
+      result
+    end
+
+    # get list of resources
+    # example: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/networkInterfaces
+    def get_resources_by_url(url, params = {})
+      result = nil
+      next_url = nil
+      begin
+        uri = http_url(url, params)
+        response = http_get(uri)
+        unless response.body.nil?
+          body = JSON.parse(response.body)
+          result = body
+          next_url = body['nextLink']
+        end
+
+	while next_url != nil
```

Comment:
minor: URI seems may throws the BadURIError. we may better to catch it too though we should trust azure api more.

Reply:
I think Azure should response a valid URL, if it isn't a valid URL, it is reasonable for URI to raise a BadURIError.